### DISCUSSION TITLE
Unable to set any of button radio group in form

### DISCUSSION THREAD

**Discussion by mgmauri** [pymupdf/PyMuPDF#2333] [Looking for help] [answered]
Hi! 
I have this form with a 4 radio button group whose widget changes its field value (from '0' to '3') when using the pdf viewer. For the 4 buttons I only see one widget
But if I try to set it using the sequence below
```
widget.field_value='1'  #or '2' or '3' or '4'
widget.update()
document.save(filename)
```
I don't see any change on any buttons state.
I know PyMuPDF does not currently support groups of (interconnected) buttons, but the [docs](https://pymupdf.readthedocs.io/en/latest/widget.html#:~:text=.%20The%20widget%20object%20also%20does%20not%20reflect%20the%20presence%20of%20a%20button%20group.%20Setting%20or%20unsetting%20happens%20via%20values%20True%20and%20False%20and%20will%20always%20work%20without%20affecting%20other%20radio%20buttons.) says that I'd still be able to change their states. Is it correct? Any ideas how could I do that?

**Comment by JorjMcKie** [comment]
Can you please let me have the example file?

**Reply by mgmauri** [reply]
Sure. Here it goes: [i-9-spanish.pdf](https://github.com/pymupdf/PyMuPDF/files/11200650/i-9-spanish.pdf)

The widget `field_name` is `form1[0].section1Page1[0].subSection1PositionWrapper[0].subSection1Bottom[0].subCitizenshipStatus[0].radioButtonListCitizenship[0]` 
and its xref is `638` (assuming that it's fix for each file)

**Reply by JorjMcKie** [reply]
The lack of support for radio button groups can be circumvented by one more instruction on your behalf.
For this field type, the value is actually a PDF **_name_** object - meaning that it starts with a slash "/". So for selecting the 4th field, you should give it a value of "/3".

```python
w = page.load_widget(638)
# initially we have an Off state:
print(doc.xref_object(638))
<<
  /AP <<
    /D <<
      /3 588 0 R
      /Off 589 0 R
    >>
    /N <<
      /3 587 0 R
    >>
  >>
  /AS /Off  % <== current state
  /DA (/ZaDb 8.00 Tf 0 g)
  /F 4
  /MK <<
    /CA (8)
  >>
  /NM (form1[0].section1Page1[0].subSection1PositionWrapper[0].subSection1Bottom[0].subCitizenshipStatus[0].radioButtonListCitizenship[0].rbAlien[0])
  /P 577 0 R
  /Parent 127 0 R
  /Rect [ 39.572 410.41 49.572 420.41 ]
  /StructParent 17
  /Subtype /Widget
  /Type /Annot
>>
# for setting this button as selected, we must do both of the following:
w.field_value = "/3"
w.update()
doc.xref_set_key(wxref, "AS", "/3")  # this is required in addition!
```
Then you should see it correctly in a PDF viewer. Unfortunately though, displaying the field value in Python will not reflect this correctly. This is a bug which will be fixed.

**Reply by mgmauri** [reply]
Thank you for your quick response Jorj. 
I wasn't able to get to see the set value on the pdf viewer though. I tried exactly as you suggested, saving a new document right after the last `x_ref_set_key call`. Is there any extra step required?
I'm using pymupdf 1.21.1 version

**Reply by JorjMcKie** [reply]
As I wrote: this is a bug, or rather not yet implemented.
As an intermediate solution, for radio buttons use `doc.xref_get_key(widget.xref, "AS")` as the value.

**Comment by JorjMcKie** [comment]
Update:
I improved the updating mechanism for radio buttons. Now everything necessary happens by setting the `field_value` to e.g. "3" and then `widget.update()`. All the rest is taken care off.
This fix will be published with the next version.

Another thing I would like to do is further simplifying this: it should be sufficient to just do `widget.field_value = True`. The information that this value has to be translated to "3" is available internally and thus can be looked up.
BTW you also can look it up: `widget.button_states()` already today returns a dictionary like this one:
```python
import fitz
doc=fitz.open("i-9-spanish.pdf")
page=doc[0]
w = page.load_widget(638)
w.button_states()
{'normal': ['3'], 'down': ['3', 'Off']}
```
There two keys: "normal" and "down". Each value is a list of either length 1 or 2. If a list item is not "Off", then it is the value for the "On" state. This value is the one you must use as per today to set the on / True.

**Reply by JorjMcKie** [reply]
Maybe I should simply make a new method like `widget.on_state()` which returns "3". Then you could use `widget.field_value = widget.on_state()`. Or similar ...

**Reply by JorjMcKie** [reply]
Done: For your example file, the method `widget.on_state()` would show this:
```python
for w in page.widgets():
    print(w.xref, w.field_type_string, "ON-value:", w.on_state())

622 Text ON-value: None
623 Text ON-value: None
624 Text ON-value: None
625 Text ON-value: None
626 Text ON-value: None
627 Text ON-value: None
628 Text ON-value: None
629 ComboBox ON-value: None
630 Text ON-value: None
631 Text ON-value: None
632 Text ON-value: None
633 Text ON-value: None
634 Text ON-value: None
635 RadioButton ON-value: 0
636 RadioButton ON-value: 1
637 RadioButton ON-value: 2
638 RadioButton ON-value: 3
639 Text ON-value: None
640 Text ON-value: None
641 Text ON-value: None
642 Text ON-value: None
643 Text ON-value: None
644 ComboBox ON-value: None
645 RadioButton ON-value: 0
646 RadioButton ON-value: 1
647 Text ON-value: None
648 Text ON-value: None
649 Text ON-value: None
650 Text ON-value: None
651 Text ON-value: None
652 ComboBox ON-value: None
653 Text ON-value: None
```
And further:
```python
w = page.load_widget(638)
w.field_value = w.on_state()
w.update()
for w in page.widgets(types=[fitz.PDF_WIDGET_TYPE_RADIOBUTTON]):
    print(w.xref, w.field_value)

    
635 Off
636 Off
637 Off
638 3
645 Off
646 Off
# checking if radiobtton is selected:
w = page.load_widget(635)
w.field_value == w.on_state()
False
w = page.load_widget(638)
w.field_value == w.on_state()
True
```

**Comment by julian-smith-artifex-com** [comment]
This is fixed in new release PyMuPDF-1.22.3.

**Comment by horrokotoj** [comment]
Hi,
I seem to have a similar issue.

I am running this and xref 716 works fine but on xref 82 I get an issue.

The problem sees to be ```field.update()```according to the fault message.

```
pdf_document = fitz.open("hyreskontrakt-för-garage-eller-garageplats.pdf")
    
for page_num in range(pdf_document.page_count):
    page = pdf_document[page_num]
    
    

    for field in page.widgets():
    print(field.xref, field.field_type_string, "ON-value:", field.on_state())

    #enligt avi
    if field.xref == 716:
        field.field_value = field.on_state()
        field.update()

    #elström
    if field.xref == 82:
        field.field_value = field.on_state()
        field.update()

# Save the modified PDF
pdf_document.save(output)
pdf_document.close()
```
```
714 RadioButton ON-value: kalenderkvartals#20b#F6rjan
715 RadioButton ON-value: kalenderm#E5nads#20b#F6rjan
716 RadioButton ON-value: enligt#20avi
79 Text ON-value: None
83 Text ON-value: None
81 RadioButton ON-value: ing#E5r
80 RadioButton ON-value: m#E4tare
84 RadioButton ON-value: ber#E4knas
82 RadioButton ON-value: debiteras#20enligt#20angivet#20s#E4tt
Traceback (most recent call last):
  File "/Users/user/Documents/pyhton_script_avtal/pymupdf_garageplats.py", line 169, in <module>
    generate(f"output2.pdf", "user", "redacted", "9028", "7200", "2024-11-01", "2024-12-31", "0", "0", "redacted", "")
  File "/Users/user/Documents/pyhton_script_avtal/pymupdf_garageplats.py", line 152, in generate
    field.update()
  File "/Users/user/.pyenv/versions/3.12.1/lib/python3.12/site-packages/fitz/__init__.py", line 7088, in update
    TOOLS._save_widget(self._annot, self)
  File "/Users/user/.pyenv/versions/3.12.1/lib/python3.12/site-packages/fitz/__init__.py", line 21443, in _save_widget
    JM_set_widget_properties(annot, widget)
  File "/Users/user/.pyenv/versions/3.12.1/lib/python3.12/site-packages/fitz/__init__.py", line 17594, in JM_set_widget_properties
    mupdf.pdf_set_field_value(pdf, annot_obj, on, 1)
  File "/Users/user/.pyenv/versions/3.12.1/lib/python3.12/site-packages/fitz/mupdf.py", line 47556, in pdf_set_field_value
    return _mupdf.pdf_set_field_value(doc, field, text, ignore_trigger_events)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: in method 'pdf_set_field_value', argument 3 of type 'char const *'
```
File I used:
[hyreskontrakt-för-garage-eller-garageplats.pdf](https://github.com/pymupdf/PyMuPDF/files/14075864/hyreskontrakt-for-garage-eller-garageplats.pdf)

Any suggestions on how I might fix this?

**Comment by ag-gaphp** [comment]
I'm also still having an issue with radio buttons. I have 2 in a document, and I can change almost any other property of the radio buttons without issue and see those changes reflected after a `doc.save(path, garbage=1)`. The only change I am unable to affect is the value of the radios.

System info:
* PyMuPDF v1.23.26
* Python v3.12.2
* Windows 10 Pro
* Using a venv in VSCodium

I have 2 radio buttons in a group, the only 2 on the page. Here is their info according to PyMuPDF:

```python
>>> fdoc = fitz.open(doc_path)
>>> page = fdoc.load_page(0)
>>> btn1 = page.load_widget(66)
>>> btn2 = page.load_widget(69)
>>> btn1.button_states()
{'normal': ['Off', '1'], 'down': None}
>>> btn2.button_states()
{'normal': ['Off', '2'], 'down': None}
>>> print(btn1.field_value, ", ", btn2.field_value)
Off, 2
```

This is how they read when I first export from LibreOffice Draw. Then I do the following:

```python
>>> btn2.field_value = 'Off'
>>> btn2.update()
>>> btn2.field_value
'Off'
```

It appears to be off. However, when I save the file and open in any PDF viewer, `btn2` is _always_ selected. This is whether I set both values to `Off`, `False`, `True`, anything. No matter the combination I've tried to turn both of them off, or at least check `btn1` instead of `btn2`, it never actually changes in the saved PDF even though PyMuPDF is reporting that the value has changed just before saving.

Attaching the PDF in question as it has been exported from LibreOffice, before I run my PyMuPDF script on it:
[ROI - General.pdf](https://github.com/pymupdf/PyMuPDF/files/14774965/ROI.-.General.pdf)

PDF viewers I've tested with: Firefox, Chrome, Edge, LaTeX/PDF viewer extenson in VSCodium, Adobe Acrobat Pro, Okular

**Reply by ag-gaphp** [reply]
I've implemented everything into my script and now I get a slightly different result. In my script I am setting any and all Radios (type 5) or Checkboxes (type 2) to a value of `Off`.

Now, my final PDF shows both RadioButtons as filled/checked in all editors, and even though I am marking them as required, they are not actually being set as required.

This issue doesn't halt my progress in my system, but I would like to get to the bottom of it at some point. I'm not entirely sure if this is a bug of some type in PyMuPDF or not. I saw some other issue threads that mention bugs that sound very similar to this have been fixed in much earlier versions of PyMuPDF (namely #711)

**Reply by ag-gaphp** [reply]
Posted as new bug #3314

**Reply by JorjMcKie** [reply]
This works for me. Note: using True/False should be sufficient.
```python
import fitz

doc = fitz.open("ROI-General.pdf")
page = doc[0]
btn1 = page.load_widget(66)
btn2 = page.load_widget(69)
print(f"btn1={btn1.field_value}, btn2={btn2.field_value} - {doc.name=}")

btn1.field_value = False
btn1.update()

btn2.field_value = True
btn2.update()

doc.save("x.pdf")

doc = fitz.open("x.pdf")
page = doc[0]
btn1 = page.load_widget(66)
btn2 = page.load_widget(69)
print(f"btn1={btn1.field_value}, btn2={btn2.field_value} - {doc.name=}")
```
Output:
```bash
btn1=1, btn2=Off - doc.name='ROI-General.pdf'
btn1=Off, btn2=2 - doc.name='x.pdf'
```

**Reply by ag-gaphp** [reply, marked as answer]
Strange. I've been at this for a few hours and have been unsuccessful in getting those radios to both be in the off state. When I use the PDF that I uploaded initially here, I am still unable to get the results I expect. In fact, when I run the script it tells me that `btn2` value is `2` and `btn1` value is `Off`. I don't know why that discrepancy is there with seemingly the same file...

However, I just made some edits to my source ODG file, and while running my script everything is working now.

The only times I get a failure now are:
1. If I use my original file, but this seems to maybe be something on my environment's end/corrupt file?
2. If I do `btn1.field_value = "Off"` instead of `False`

Setting the value to `False` specifically appears to be working when I run my python process from start to finish now, though. I'm trying to locate the exact spot in the documentation that led me to believe setting `'Off'` was the correct way to go (because it appears in the `button_states` list), but it may have been a reply you made within a discussion/issue that I am remembering and got confused.

**Reply by JorjMcKie** [reply]
No worries.
As I wrote it should suffice to use `False`/`True` for buttons.

Bottom line, you should be able to achieve what you want.

**Reply by ag-gaphp** [reply]
Definitely, all is working well now with that adjustment. I appreciate it!
